lib/common/crmcommon_private.h

```diff
@@ -103,6 +103,84 @@ pcmk__xml_attr_value(const xmlAttr *attr)
 
 #define PCMK__IPC_VERSION 1
 
+// IPC behavior that varies by daemon
+typedef struct pcmk__ipc_methods_s {
+    /*!
+     * \internal
+     * \brief Allocate any private data needed by daemon IPC
+     *
+     * \param[in] api  IPC API connection
+     *
+     * \return Standard Pacemaker return code
+     */
+    int (*new_data)(pcmk_ipc_api_t *api);
+
+    /*!
+     * \internal
+     * \brief Free any private data used by daemon IPC
+     *
+     * \param[in] api_data  Data allocated by new_data() method
+     */
+    void (*free_data)(void *api_data);
+
+    /*!
+     * \internal
+     * \brief Perform daemon-specific handling after successful connection
+     *
+     * Some daemons require clients to register before sending any other
+     * commands. The controller requires a CRM_OP_HELLO (with no reply), and
+     * the CIB manager, executor, and fencer require a CRM_OP_REGISTER (with a
+     * reply). Ideally this would be consistent across all daemons, but for now
+     * this allows each to do its own authorization.
+     *
+     * \param[in] api  IPC API connection
+     *
+     * \return Standard Pacemaker return code
+     */
+    int (*post_connect)(pcmk_ipc_api_t *api);
+
+    /*!
+     * \internal
+     * \brief Check whether an IPC request results in a reply
+     *
+     * \parma[in] api      IPC API connection
+     * \param[in] request  IPC request XML
+     *
+     * \return true if request would result in an IPC reply, false otherwise
+     */
+    bool (*reply_expected)(pcmk_ipc_api_t *api, xmlNode *request);
+
+    /*!
+     * \internal
+     * \brief Perform daemon-specific handling of an IPC message
+     *
+     * \param[in] api  IPC API connection
+     * \param[in] msg  Message read from IPC connection
+     */
+    void (*dispatch)(pcmk_ipc_api_t *api, xmlNode *msg);
+
+    /*!
+     * \internal
+     * \brief Perform daemon-specific handling of an IPC disconnect
+     *
+     * \param[in] api  IPC API connection
+     */
+    void (*post_disconnect)(pcmk_ipc_api_t *api);
+} pcmk__ipc_methods_t;
+
+// Implementation of pcmk_ipc_api_t
+struct pcmk_ipc_api_s {
+    enum pcmk_ipc_server server;          // Daemon this IPC API instance is for
+    enum pcmk_ipc_dispatch dispatch_type; // How replies should be dispatched
+    crm_ipc_t *ipc;                       // IPC connection
+    mainloop_io_t *mainloop_io;     // If using mainloop, I/O source for IPC
+    bool free_on_disconnect;        // Whether disconnect should free object
```

Comment:
This might work better as part of a bit field so we're a little more future-proof in this public API.

Reply:
This part is private API, we can be flexible